db.go

```diff
@@ -188,6 +188,25 @@ func Open(dir string, l log.Logger, r prometheus.Registerer, opts *Options) (db
 	return db, nil
 }
 
+// QueryLatest will return the matching series with their latest value.
+// Useful for federation.
+func (db *DB) QueryLatest(cutoff int64, matchers ...labels.Matcher) []SeriesLatest {
+	db.headmtx.RLock()
+	defer db.headmtx.RUnlock()
+
+	sl := make([]SeriesLatest, 0)
+	for _, h := range db.heads {
+		m := h.Meta()
+		if m.MaxTime < cutoff {
+			continue
+		}
+
+		sl = append(sl, h.SelectLatest(cutoff, matchers...)...)
+	}
```

Comment:
This logically doesn't seem correct.

The `heads` list is _not_ sharded across all our series but represents blocks sequential in time, each covering the entire series space.
What we do here is checking every head for the last sample, which generally means we get several samples for the same series, and throwing them all into a big result.

We have to first consider only heads with `m.MaxTime >= cutoff` as you correctly do here.
But then we have to start from the last block, look which series exist there, and take their last sample.
Then go to the block before that and do the same, but only add those series and samples to the result, which are not already in our result so far. Repeat for all considered blocks.

Going with slices here is fine for sketching things out, but we will have to make this all iterator based.
Especially if we want to "merge with priority" across blocks, with sorted iterators we can do so in O(n) whereas the naive approach will be quadratic, cubic, etc... depending on how many heads have to be considered.